### PR TITLE
Support complete sharded GGUF imports

### PR DESCRIPTION
**Pull request by justinchuby**
## Summary
- assemble a complete local GGUF shard set as one lazy logical model while retaining every reader/memmap
- validate filename identity, contiguous `split.no`, matching `split.count` and `split.tensors.count`, compatible metadata/container properties, tensor extents, and unique tensor names
- resolve Hugging Face shard references as an exact complete set pinned to one immutable commit, with cache-aware total-size/free-space preflight before downloads
- preserve single-file behavior, quantized tensor import, and runtime artifact identity hashing across Hub snapshot symlinks
- document generated user-facing shard behavior

## Tests
- `python -m pytest src/mobius/integrations/gguf/_reader_test.py src/mobius/integrations/gguf/_shard_set_test.py src/mobius/integrations/gguf/_preflight_test.py src/mobius/integrations/gguf/_runtime_evidence_test.py src/mobius/integrations/gguf/_builder_test.py -q --tb=short` (538 passed)
- `python -m pytest tests/build_graph_test.py tests/cli_test.py src/ -q -k 'not phi4mm and not apply_weights_unknown' --tb=short` (8,362 passed, 57 skipped, 12 deselected)
- `lintrunner f --output oneline --all-files`
- `lintrunner -a`

All shard coverage uses tiny generated GGUF files and mocked Hub metadata/download boundaries. The motivating multi-gigabyte artifact was not downloaded.

### CODE REVIEW

**Comment by justinchuby** [comment]
Closing as superseded by the sharded GGUF implementation and subsequent hardening that landed through the Qwen4Exp/mainline work. The unresolved Copilot findings are being audited against current main; any still-applicable gaps will be fixed in a focused follow-up rather than merging this conflicting historical branch.

File: src/mobius/integrations/gguf/_shard_set.py [hunk 1]
```diff
@@ -533,6 +594,16 @@ def _validate_shard_set(infos: list[ShardInfo], shards: list[GGUFModel]) -> None
...
+    prefixes = {
+        parsed[0]
+        for info in infos
+        if (parsed := parse_shard_filename(info.path.name)) is not None
+    }
+    if len(prefixes) != 1:
+        raise GgufShardError(
+            f"Shard filenames have inconsistent basenames {sorted(prefixes)}; "
+            "the files belong to different split sets."
+        )
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/integrations/gguf/_shard_set.py:606]: Superseded on current `main` by the sharded implementation merged through #656. `_build_shard_infos()` now falls back to authoritative `split.no`/`split.count` metadata for noncanonical filenames, while `_validate_shard_set()` enforces basename consistency only for filenames that parse. Renamed split sets also have bounded-header discovery tests, so this no longer produces the misleading empty-prefix error.

File: src/mobius/integrations/gguf/_shard_set.py [hunk 2]
```diff
@@ -619,6 +718,45 @@ def _validate_identity(infos: list[ShardInfo], shards: list[GGUFModel]) -> None:
...
+        for key, value in shard.metadata.items():
+            if key == _SPLIT_NO:
+                continue
+            # gguf-py exposes container-local pseudo fields alongside serialized
+            # metadata. Keep the primary values, then canonicalize tensor_count
+            # below for the logical model.
+            if key.startswith("GGUF."):
+                if key not in merged:
+                    merged[key] = value
+                    owners[key] = info.path.name
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/integrations/gguf/_shard_set.py:735]: Superseded on current `main` by the sharded implementation merged through #656. `_merge_metadata()` now ignores continuation-shard `GGUF.*` pseudo fields, retains the primary container-local values, and explicitly canonicalizes `GGUF.tensor_count` from the complete validated shard set. Container version and byte order are exposed and validated separately.

File: src/mobius/integrations/gguf/_builder.py [hunk 3]
```diff
@@ -4905,7 +5116,7 @@ def build_from_gguf(
     gguf_path = _resolve_gguf_path(gguf_path)
     logical_source_filename = _logical_source_filename(source_reference, gguf_path)
     source_path = Path(gguf_path)
-    if source_path.is_symlink():
+    if source_path.is_symlink() and _GGUF_SHARD_FILENAME_RE.search(source_path.name) is None:
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/integrations/gguf/_builder.py:5119]: Addressed in follow-up #675 against current `main`. Sharded builds now resolve any opened local symlink path to its regular-file target and bind it through `GgufShardSet._set_identity_paths()`, which verifies `samefile()` before evidence hashing. The digest still frames the logical shard filenames. Regression coverage exercises `build_from_gguf()` with a local symlinked split set outside `HF_HUB_CACHE` and successfully computes the artifact identity.

File: src/mobius/integrations/gguf/_builder.py [hunk 4]
```diff
@@ -4917,6 +5128,10 @@ def build_from_gguf(
             source_path = source_path.resolve(strict=True)
             gguf_path = str(source_path)
     gguf_model = _gguf_model if _gguf_model is not None else open_gguf_model(gguf_path)
+    if isinstance(gguf_model, GgufShardSet):
+        identity_paths = _hub_cache_identity_paths(gguf_model.shard_paths)
+        if identity_paths is not None:
+            gguf_model._set_identity_paths(identity_paths)
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/integrations/gguf/_builder.py:5134]: Addressed in follow-up #675 against current `main`. Sharded builds now resolve any opened local symlink path to its regular-file target and bind it through `GgufShardSet._set_identity_paths()`, which verifies `samefile()` before evidence hashing. The digest still frames the logical shard filenames. Regression coverage exercises `build_from_gguf()` with a local symlinked split set outside `HF_HUB_CACHE` and successfully computes the artifact identity.

File: src/mobius/integrations/gguf/_shard_set.py [hunk 5]
```diff
@@ -545,35 +616,49 @@ def _validate_shard_set(infos: list[ShardInfo], shards: list[GGUFModel]) -> None
         )
 
     # Authoritative split.count metadata must agree with the filename count.
-    declared_counts = {info.split_count for info in infos if info.split_count is not None}
-    if declared_counts and declared_counts != {filename_count}:
+    missing_counts = [info.path.name for info in infos if info.split_count is None]
+    if missing_counts:
+        raise GgufShardError(
+            f"Every shard must declare split.count; missing from {missing_counts}."
+        )
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/integrations/gguf/_shard_set.py:623]: Addressed in follow-up #675 against current `main`. Missing `split.count`, `split.no`, and `split.tensors.count` diagnostics now sort and quote filenames before formatting. A parameterized regression supplies reversed shard order and checks the exact deterministic message for all three fields.
